llmfoundry/models/llm_embed/modeling_llm_embed.py

```diff
@@ -209,23 +208,38 @@ def construct_model(self):
             self.is_mpt = True
         return model
 
+    def _update_step_size_if_needed(self, batch: MutableMapping) -> None:
+        """Update step size on first batch if we detect hard negatives."""
+        if self.step_size:
+            return
+
+        input_shape = batch['input_ids'].shape
+        if input_shape[1] > 2:
+            # We have hard negatives, batch shape is [batch, sample of query+positive passage+negative passages, tokens].
+            self.step_size = input_shape[1]
+            log.info(
+                f'Detected hard negatives, updated step_size to {self.step_size}',
+            )
+            self.gather_in_batch_negatives = False
```

Comment:
Oh, hmm.

Ideally, we could do in-batch negatives AND hard negatives at the same time.

IDK if the code supported this before, and suggest we simply retain the existing behavior. In either case, in-batch negatives is something that should be configurable IMO.

Reply:
Thats reasonable